src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int Count<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+            
+            IEnumerator<T> enumerator = source.GetEnumerator();
+            return CountAndDispose(enumerator, enumerator);
+        }
+
+        /// <summary>
+        /// Gets the number of items in the enumerator, and then disposes it.
+        /// </summary>
+        /// <param name="enumerator">The enumerator to count.</param>
+        /// <param name="disposable">The enumerator typed as an <see cref="IDisposable"/>.</param>
+        /// <returns>The count of the enumerator.</returns>
+        /// <remarks>This method should be kept non-generic to avoid generated code bloat.</remarks>
```

Comment:
@stephentoub I did so because I figured 1 or 2 extra procedure calls wouldn't be much compared to the overhead of repeatedly calling `MoveNext`, a bunch of typecasts beforehand, `GetEnumerator` / `Dispose`, etc. I did a small benchmark [here](https://gist.github.com/jamesqo/22330e09b10f1f66c83a86fb3705b7a0) testing for regression; surprisingly, the difference does not seem to be measurable even for the length 1 case.